src/router.ts

```diff
@@ -30,13 +30,32 @@ export const IS_PATTERN = /[*:{}+?()]/;
 
 export class UrlPatternRouter<T> implements Router<T> {
   readonly _routes: Route<T>[] = [];
-  readonly _middlewares: T[] = [];
+  readonly _middlewares: Route<T>[] = [];
 
-  addMiddleware(fn: T): void {
-    this._middlewares.push(fn);
+  addMiddleware(pathname: string | URLPattern, handler: T): void {
+    if (
+      typeof pathname === "string" && pathname !== "/*" &&
+      IS_PATTERN.test(pathname)
+    ) {
+      this._middlewares.push({
+        path: new URLPattern({ pathname }),
+        handlers: [handler],
+        method: "MIDDLEWARE",
+      });
+    } else {
+      this._middlewares.push({
+        path: pathname,
+        handlers: [handler],
+        method: "MIDDLEWARE",
+      });
+    }
```

Comment:
```suggestion
    this._middlewares.push({
      path: isPattern(pathname) ? new URLPattern({ pathname }) : pathname,
      handlers: [handler],
      method: "MIDDLEWARE",
    });
```
We can factor out that conditional into an `isPattern()` which can be reused on lines 60-61.

Reply:
Thanks!
I extracted it as isURLPattern.